org.alloytools.alloy.classic.provider/src/main/java/org/alloytools/alloy/classic/solver/kodkod/AbstractKodkodSolver.java

```diff
@@ -118,27 +118,23 @@ public ITupleSet none() {
             @Override
             public Iterator<AlloyInstance> iterator() {
                 return new Iterator<AlloyInstance>() {
-
+                    //This iterator delegates to the a4 iterator
+                    //We'll have a next solution when a4's next solution is satisfiable
                     A4Solution nextSolution = ai;
-                    boolean	first= true;
-                    
+
                     @Override
                     public boolean hasNext() {
                         return nextSolution.satisfiable();
                     }
 
                     @Override
                     public AlloyInstance next() {
-                    	
                     	if ( !hasNext())
                     		throw new NoSuchElementException("No instance available");
                     	
-                    	A4Solution solution = this.nextSolution;
-                    	if ( !first) {
-                    		this.nextSolution = solution.next();
-                    	} else
-                    		first = false;
-                    	
+                    	A4Solution solution = nextSolution;
+                        nextSolution = nextSolution.next();
```

Comment:
Correct me if I'm wrong, but this means that when asked to return the current solution this iterator will eagerly solve for next before returning the current one.  This in turn means that the user has to wait twice as long before it gets the first solution, which would be is unacceptable (knowing that in many cases it takes minutes or even hours to find a solution)

Reply:
That's a great point, I didn't consider the performance implications when I made this change --- I was just trying to make the iterator correct.

From what I can tell, a solution needs to be generated before we know whether it's satisfiable, so we may need to generate the next solution as part of `hasNext`, and cache that result to return when `next` is actually called.

That behavior (`hasNext` is slow because it's generating a full solution) may be surprising, but I don't know if it can be avoided.

Reply:
Yes, it would make more sense to search for a new solution in `hasNext`.